File: Easy/145. Binary Tree Postorder Traversal/solution.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<int> PostorderTraversal(TreeNode root) {
        List<int> result = new List<int> ();
        
        Stack<TreeNode> myStack = new Stack<TreeNode> ();
        TreeNode currentNode = root;
        
        while (myStack.Count > 0 || currentNode != null){
            if (currentNode != null){
                myStack.Push(currentNode);
                currentNode = currentNode.left;
            }
            
            else{
                TreeNode temp = myStack.Peek().right;
                if (temp == null){
                    temp = myStack.Pop();
                    result.Add(temp.val);
                    while (myStack.Count > 0 && myStack.Peek().right == temp){
                        temp = myStack.Pop();
                        result.Add(temp.val);
                    }
                }
                else{
                    currentNode = temp;
                }
            }
        }
        
        return result;
    }
}

File: Easy/145. Binary Tree Postorder Traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        stack<TreeNode*> myStack;
        
        vector<int> result;
        
        TreeNode* curNode = root;
        
        while (curNode || !myStack.empty()){
            if (curNode){
                myStack.push(curNode);
                curNode = curNode -> left;
            }
            
            else{
                TreeNode* tempNode = myStack.top();
                
                if (tempNode -> right){
                    curNode = tempNode -> right;
                }
                
                else{
                    result.push_back(tempNode -> val);
                    tempNode = myStack.top();
                    myStack.pop();
                    
                    while(!myStack.empty() && tempNode == myStack.top() -> right){
                        tempNode = myStack.top();
                        result.push_back(tempNode -> val);
                        myStack.pop();
                    }
                }
            }
        }
        
        return result;
    }
};